tool-openssl/req.cc

```diff
@@ -0,0 +1,675 @@
+    return nullptr;
+  }
+
+  std::string type;
+  std::string value;
+
+  while (*subject_name_ptr) {
+    // Reset strings for new iteration
+    type.clear();
+    value.clear();
+
+    // Parse type
+    while (*subject_name_ptr && *subject_name_ptr != '=') {
+      type += *subject_name_ptr++;
+    }
+
+    if (!*subject_name_ptr) {
+      fprintf(stderr, "Hit end of string before finding the equals.\n");
+      return nullptr;
+    }
+
+    // Skip '='
+    subject_name_ptr++;
+
+    // Parse value
+    while (*subject_name_ptr && *subject_name_ptr != '/') {
+      if (*subject_name_ptr == '\\' && *(subject_name_ptr + 1)) {
+        // Handle escaped character
+        subject_name_ptr++;
+        value += *subject_name_ptr++;
+      } else {
+        value += *subject_name_ptr++;
+      }
+    }
+
+    // Skip trailing '/' if present
+    if (*subject_name_ptr == '/') {
+      subject_name_ptr++;
+    }
+
+    // Convert type to NID, skip unknown attributes
+    int nid = OBJ_txt2nid(type.c_str());
+    if (nid == NID_undef) {
+      fprintf(stderr, "Warning: Skipping unknown attribute \"%s\"\n", type.c_str());
+      // Skip unknown attributes
+      continue;
+    }
+
+    // Skip empty values
+    if (value.empty()) {
+      fprintf(stderr, "Warning: No value specified for attribute \"%s\", Skipped\n", type.c_str());
+      continue;
+    }
+
+    // Add entry to the name
+    if (!X509_NAME_add_entry_by_NID(name.get(), nid, DEFAULT_CHAR_TYPE,
+                                   (unsigned char*)value.c_str(),
+                                   -1, -1, 0)) {
+      OPENSSL_PUT_ERROR(X509, ERR_R_X509_LIB);
+      return nullptr;
+    }
+  }
+
+  return name.release();
+}
+
+static X509_NAME *prompt_for_subject(X509_REQ* req, bool isCSR, unsigned long chtype = MBSTRING_ASC) {
+  // Default values for subject fields
+  const struct {
+    const char* field_name;
+    const char* short_desc;
+    const char* default_value;
+    int nid;
+  } subject_fields[] = {
+    {"countryName", "Country Name (2 letter code)", "AU", NID_countryName},
+    {"stateOrProvinceName", "State or Province Name (full name)", "Some-State", NID_stateOrProvinceName},
+    {"localityName", "Locality Name (eg, city)", "", NID_localityName},
+    {"organizationName", "Organization Name (eg, company)", "Internet Widgits Pty Ltd", NID_organizationName},
+    {"organizationalUnitName", "Organizational Unit Name (eg, section)", "", NID_organizationalUnitName},
+    {"commonName", "Common Name (e.g. server FQDN or YOUR name)", "", NID_commonName},
+    {"emailAddress", "Email Address", "", NID_pkcs9_emailAddress}
+  };
+
+  // Extra attributes for CSR
+  const struct {
+    const char* field_name;
+    const char* short_desc;
+    const char* default_value;
+    int nid;
+  } extra_attributes[] = {
+    {"challengePassword", "A challenge password", "", NID_pkcs9_challengePassword},
+    {"unstructuredName", "An optional company name", "", NID_pkcs9_unstructuredName}
+  };
```

Comment:
NP: Since this struct is used twice, we could give it a name and declare it outside the function.